### ISSUE TITLE
Support for equality constraints

### ISSUE THREAD

**Issue by redcican** [PKU-DAIR/open-box#31]
请问你们支持恒等约束吗? 如果接受，该如何设置?

**Comment by jhj0411jhj** [comment]
我认为在黑盒优化中，优化**黑盒恒等约束**是十分困难的。目前OpenBox不支持黑盒恒等约束。您可以尝试将恒等约束 $c(x) = 0$ 转化为两个不等式约束 $c(x)-\epsilon \le 0$ 和 $c(x)+\epsilon \ge 0$，然后执行优化。但是，这样做的效果不被保证，并且恒等约束不会被100%满足。

最近，我们实现了一个新的功能以支持**对输入超参数进行约束** (https://github.com/PKU-DAIR/open-box/commit/be455c281ef48c9a7ad8a2f811497a8f1e96e15d) ，并发布了一个新的版本 (`pip install openbox>=0.7.18`)。如果您的约束是关于输入超参数的，请参考(#37)以获取更多信息。

此外，我想知道您的恒等约束是否是类似这样的形式：`a+b+c=0`？其中`a,b,c`是超参数。在这种情况下，我建议您只定义两个超参数`a`和`b`，然后在目标函数中计算`c=-a-b`。

-----

I think in black-box optimization, it's very hard to optimize **black-box equality constraints**. Currently OpenBox has no support for black-box equality constraints. You can try to convert the equality constraint $c(x) = 0$ into two inequality constraints $c(x)-\epsilon \le 0$ and $c(x)+\epsilon \ge 0$, and then run the optimization. However, the performance is not guarantted and the equality constraint cannot be 100% satisfied.

Recently, we added a new feature to support **constraints (conditions) between hyperparameters** (https://github.com/PKU-DAIR/open-box/commit/be455c281ef48c9a7ad8a2f811497a8f1e96e15d) and released a new version (`pip install openbox>=0.7.18`). If your constraint is about input hyperparameters, please refer to this issue (#37) for more details.

Furthermore, I would like to know if your equality constraint is like `a+b+c=0`, where `a,b,c` are hyperparameters. In this situation, I suggest you define two parameters `a` and `b`, and calculate `c=-a-b` in the objective function.

**Comment by jhj0411jhj** [comment]
If the constraint is like $x_1+x_2+x_3=m$, and you want $x_i \in [0,1]$, I suggest you define two hyperparameters $x_1$ and $x_2$, and define a `sample_condition`, which calculates $x_3=m-x_1-x_2$, and judges whether $x_3 \in [0,1]$:
```python
def sample_condition(config):
    x3 = m - config['x1'] - config['x2']
    return 0 <= x3 <= 1
```
